### ISSUE TITLE
SPV remapper: does not handle wide literals, e.g. OpSwitch on 64-bit type

### ISSUE THREAD

**Issue by dneto0** [KhronosGroup/glslang#435]
The remapper doesn't have the machinery yet to handle literals wider than 1 word.  This becomes a problem when trying to track Ids used by an OpSwitch over a 64-bit value.

Example:
```
               OpCapability Shader
               OpMemoryModel Logical GLSL450
       %void = OpTypeVoid
       %long = OpTypeInt 64 1
          %3 = OpTypeFunction %void
          %4 = OpConstant %long 0
          %5 = OpFunction %void None %3
          %6 = OpLabel
               OpSelectionMerge %7 None
               OpSwitch %4 %7 0 %8 1 %9
          %8 = OpLabel
               OpBranch %7
          %9 = OpLabel
               OpBranch %7
          %7 = OpLabel
               OpReturn
               OpFunctionEnd
```

I get this internal error:
spirv-remap: SPVRemapper.cpp:219: spv::Id spv::spirvbin_t::localId(spv::Id, spv::Id): Assertion `id != spv::NoResult && newId != spv::NoResult' failed.

I suggest filing this as an enhancement, except that it results in an internal error. ?

**Comment by steve-lunarg** [comment]
I can investigate this. 

@dneto0: do you have a small GLSL shader I can use to duplicate it?  I tried making this one similar to your SPIR-V, but receive an error that switch conditions must be a scalar integers.  If I cast 'x' to a plain old 32 bit int, such as switch(int(x)), then it doesn't illustrate the problem any more.
```

#version 450
#extension GL_ARB_gpu_shader_int64 : enable

void main()
{
    int64_t x = 0;

    switch (x) { case 0: break; }
}

```

**Comment by dneto0** [comment]
@steve-lunarg
I don't think it can be done in GLSL.
GLSL 4.50 section 6.3 says "The type of the init-expression value in a switch statement must be a scalar int or uint. "    And GL_ARB_gpu_shader_int64 doesn't extend the switch statement semantics.

You should be able to use the SPIRV-Tools assembler to assemble the above program.  I've tried attaching it, but GitHub doesn't understand .spv files...

**Comment by dneto0** [comment]
FYI.  The SPIRV-Tools handle bit widths up to 64.  The analog of what you'll need can be seen in its binary parser, around setting and using the entries in the type_id_to_number_type_info map.
E.g. see https://github.com/KhronosGroup/SPIRV-Tools/blob/master/source/binary.cpp#L207

**Comment by steve-lunarg** [comment]
OK, thanks.  I didn't know about that assembler.  I'll check it out.

**Comment by dneto0** [comment]
You should find SPIRV-Tools very useful: It has a disassembler and a validator too.

**Comment by steve-lunarg** [comment]
@dneto0: I pushed a WIP PR #456 which has a candidate fix for this issue.  It works on the example SPIR-V you provided above (which actually collapses to a null shader, since the function is dead making the types also dead).  Perhaps you could check it out on workloads you care about, and give it a thumbs up or down?

Interestingly, this imposed a new requirement on SPIR-V instruction & arg iteration.  Previously, local knowledge of the instruction was sufficient, although sometimes implicitly rather than explicitly sufficient (e.g, one argument of variable width inferred from total instruction width).  In this case proper iteration required knowing the type of an argument defined earlier in the module, so I added a new function for this purpose.  I have only tested it on your data set above with the OpConstant, but I believe it should work arbitrarily.

> You should find SPIRV-Tools very useful:

Yes!  I like it.  (I've only tried -as and -dis so far, but will check out -val soon).

@johnkslang has mentioned that SPIRV-Tools has a generic SPIR-V walker, and also the remapper tool has its own, which is one more walker than one might hope to have.  I haven't looked at the SPIRV-Tools code yet, but assuming the pragmatics work out, perhaps I could change the remapper to use the SPIRV-Tools walker, to avoid duplication.  That also might conceivably have avoided this defect.

**Comment by dneto0** [comment]
@steve-lunarg Your patch works for me.  I left comments on the PR.

Given that I worked on the SPIRV-Tools walker ("binary parser") I might be biased, but I think it's terrific.
For one thing, it's been fuzzed so I trust it on untrusted input.  It's the foundation under the SPIR-V validator, the disassembler, and (new this week), the control-flow-graph dumper.
It has a SAX-like callback C API, and lets you mostly focus on a higher level than bits and bytes.

In particular, the operands have been predigested, including handling of multi-word "logical operands" as I call them.

API description for the binary parser:
https://github.com/KhronosGroup/SPIRV-Tools/blob/master/include/spirv-tools/libspirv.h#L449

Example use:  
- the full disassembler (elaborate use): https://github.com/KhronosGroup/SPIRV-Tools/blob/master/source/disassemble.cpp
- control flow dumper (simple use, only pays attention to control flow): https://github.com/KhronosGroup/SPIRV-Tools/blob/master/tools/cfg/bin_to_dot.cpp

**Comment by steve-lunarg** [comment]
Fix attempt number two: #470.

**Comment by steve-lunarg** [comment]
One issue I see related to using the SPIRV-Tools binary parser in the remapper is constness.  The current interface seems set up for the callbacks to examine but not modify the SPIR-V as it's going by.  (Which is sensible; it's hard to parse robustly in the face of arbitrary in-place modifications).  Thus the spv_parsed_instruction_fn_t callback contains a pointer to const words.

Currently the remapper uses in-place modification of the SPIR-V stream.  That's not an insurmountable difference, because (AFAICS) I can incrementally build another SPIR-V blob during a parsing pass over const data with the changes I want.  It will mean a little more redesign, so not quite a drop in replacement.

(Nothing needs to change about the parser; I'm just posting the observation in case I have some fundamental misunderstanding about it).

**Comment by steve-lunarg** [comment]
I think this can be closed as of #471 (same code as in #470, where the Travis-CI Mac build never started for some reason).
